### ISSUE TITLE
amp-carousel: Incorrect focus handling

### ISSUE THREAD

**Issue by TetraLogicalHelpdesk** [ampproject/amphtml#29558]
## WCAG Level
Level A

## Priority
High

## Type
Navigation

## Pages/screens/Components affected
* [amp-carousel](https://amp.dev/documentation/components/amp-carousel/?format=websites)

## Description
For people who navigate using a keyboard or keyboard-like device (such as a switch control or a joystick) the focus order of interactive controls in a document should be logical and intuitive. In particular, when a page changes dynamically as a result of user interaction, it's important that focus is, where necessary, moved to the most appropriate place within the page.

In a non-looping `<amp-carousel>`, the previous/next buttons are not displayed when the carousel is showing the first/last item, respectively. The relevant button is hidden using CSS:

```
amp-carousel .amp-carousel-button.amp-disabled {
    animation: none;
    opacity: 0;
    visibility: hidden;
}
```

```
<div tabindex="0" class="amp-carousel-button amp-carousel-button-prev amp-disabled" role="button" title="..." aria-disabled="true"></div>
```

When the element is hidden in this way, it is effectively removed from the page - it is not keyboard-focusable anymore (despite the `tabindex="0"`), and hidden from assistive technologies/screen readers. This makes it problematic when a keyboard user operates the slide controls, as the control that they're focused on may essentially "disappear" from under their focus. Depending on the specific browser/assistive technology used, this can result in the focus being lost and reset to the start of the page.

## User impact
When focus is not programmatically moved to the most appropriate location following an interaction, users will be left to manually navigate again to the most appropriate location in the page. For sighted keyboard users, this can be an inconvenience/irritation, particularly if focus was lost/reset to the start of the document and they have to cycle through the entire page again. Blind and low vision assistive technology users may be unaware of what change actually occurred on the page as a result of their interaction. After performing an action, these users may be confused/left wondering if their action was actually successful. In cases where focus is lost (generally caused by the element that previously had focus being disabled or removed from the page), some browsers may error-correct and attempt to relocate the user's focus to the nearest focusable element, or they may simply reset the focus to the start of the document. Overall, it will likely make the user's experience of navigating the page unnecessarily tedious and confusing.

## Required solution
Instead of hiding the control using `visibility:hidden`, we recommend instead only visually hiding the disabled/inactive control (using a set of styles such as the `.visually-hidden` class outlined in [The A11Y Project - How-to: Hide content](https://www.a11yproject.com/posts/2013-01-11-how-to-hide-content/)). Alternatively, rather than making the control fully invisible, you could apply styles that make the button very faint/more transparent.

As changing the styling this way will make the disabled control still focusable, we recommend dynamically changing the `tabindex="0"` to `tabindex="-1"` - this way, it won't be in the regular focus order anymore. However, it will still be able to receive programmatic focus.

If a user is focused on one of the buttons, activates it, and it leads to the button becoming disabled, also make sure to then programmatically focus the control (which now has `tabindex="-1"`). This makes sure that focus is not lost/reset. Assistive technology users will now also hear that the control is disabled (as indicated by the `aria-disabled="true"` that is already being added to the markup, but which was never exposed since the entire button was removed from the accessibility tree due to the `visibility:hidden`).

```
amp-carousel .amp-carousel-button.amp-disabled {
    clip: rect(0 0 0 0);
    clip-path: inset(50%);
    height: 1px;
    overflow: hidden;
    position: absolute;
    white-space: nowrap;
    width: 1px;
}
```

```
<div tabindex="-1" class="amp-carousel-button amp-carousel-button-prev amp-disabled" role="button" title="..." aria-disabled="true"></div>
```

## Test procedure(s)
Use these steps to confirm that the solution has been correctly applied:
1. Use `TAB`/`SHIFT`+`TAB` to move keyboard focus to the previous/next buttons in a non-looping carousel.
2. Interact with the buttons until you reach the first or last slide.
3. Verify that when the previous or next button disappear, focus is not reset to the start of the page (this can be difficult to verify, as some browsers error-correct for these situations - the safest way to test this is to use developer tools and to check `document.activeElement` in the console - if it returns `<body>`, then focus was lost/reset).
4. If following the above recommendation of keeping the button in place, only hiding it visually, and suppressing its focusability: verify that the styles applied to the now disabled button don't contain `visibility:hidden` or `display:none` (as these remove the element from the accessibility tree/document structure), that they have `tabindex="-1"`, and that focus is still on that control after it has been disabled (as above, using `document.activeElement` and verifying that it reports the disabled button as being the active element).

## Related standards
* [WCAG 2.1 Success Criterion 2.4.3 Focus Order (Level A)](https://www.w3.org/TR/WCAG21/#focus-order)

## More information
* [W3C - Placing the interactive elements in an order that follows sequences and relationships within the content](https://www.w3.org/WAI/WCAG21/Techniques/general/G59)
* [The A11Y Project - How-to: Hide content](https://www.a11yproject.com/posts/2013-01-11-how-to-hide-content/)

## Test data
Test date: July 2020
